File: R_code/Kepler-code/MainFunctions/cross-validation.R
# By Bo Ning
# Mar. 5. 2018

# Cross validation function
#' @param data.train: input training data
#' @param data.sg.train: input measurement errors for the training data
#' @param bounds: a vector contains four elements, from left to right:
#'                the upper bound for mass, the lower bound for mass
#'                the upper bound for radius, the lower bound for radius
#' @param data.test: input test data
#' @param data.sg.test: input measurement errors for the test data
#' @param deg: degree used in Bernstein polynomials
#' @param abs.tol: tolerance number of computing

cross.validation <- function(data.train, data.sg.train = NULL, bounds, 
                             data.test, data.sg.test = NULL, 
                             deg, log = FALSE, abs.tol = 1e-20) {
  
  # fit MLE using training dataset
  weights <- MLE.fit(data = data.train, sigma = data.sg.train, 
                     bounds = bounds, deg = deg, log = log,
                     abs.tol = abs.tol, output.weights.only = TRUE)
  
  # calculate log-likelihood using test datasets based on estimated weights
  # using training data
  if (dim(as.matrix(data.test))[2] == 1) {
    n.test <- 1
    M.test <- data.test[1]
    R.test <- data.test[2]
  } else {
    n.test <- dim(data.test)[1]
    M.test <- data.test[, 1]
    R.test <- data.test[, 2]
  }
  
  # read the sigma
  if (is.null(data.sg.test) == FALSE) {
    if (dim(as.matrix(data.sg.test))[2] == 1) {
      M.sg.test <- data.sg.test[1]
      R.sg.test <- data.sg.test[2]
    } else {
      M.sg.test <- data.sg.test[, 1]
      R.sg.test <- data.sg.test[, 2]
    }
  }
    
  # specify the bounds
  M.max <- bounds[1]
  M.min <- bounds[2]
  R.max <- bounds[3]
  R.min <- bounds[4]
  
  # calculate cdf and pdf of M and R for each term
  # the first and last term is set to 0 to avoid boundary effects
  # so we only need to calculate 2:(deg^2-1) terms
  deg.vec <- 1:deg
  
  if (is.null(data.sg.test) == TRUE) {
    
    # pdf for Mass for each beta density
    M.indv.pdf <- as.matrix(
      sapply(deg.vec, function(x, degree) {dbeta(x, degree, deg-degree+1)},
             x = (M.test-M.min)/(M.max-M.min)), n.test, deg) / (M.max-M.min)
    
    # pdf for Radius for each beta density
    R.indv.pdf <- as.matrix(
      sapply(deg.vec, function(x, degree) {dbeta(x, degree, deg-degree+1)},
             x = (R.test-R.min)/(R.max-R.min)), n.test, deg) / (R.max-R.min)
    
    
  } else {
    
    # pdf for Mass for integrated beta density and normal density
    res.M.pdf <-  
      sapply(deg.vec, 
             function(data, data.sg, deg, degree, x.max, x.min, 
                      log = FALSE, abs.tol = 1e-10){
               apply(cbind(data, data.sg), 1, fun.for.integrate,
                     deg = deg, degree = degree, x.max = x.max, x.min = x.min,
                     log = log, abs.tol = abs.tol)},
             data = M.test, data.sg = M.sg.test, deg = deg, x.max = M.max, x.min = M.min, 
             log = log, abs.tol = abs.tol)
    M.indv.pdf <- as.matrix(res.M.pdf, n.test, deg)
    
    # pdf for Radius for integrated beta density and normal density
    res.R.pdf <-  
      sapply(deg.vec, 
             function(data, data.sg, deg, degree, x.max, x.min, 
                      log = FALSE, abs.tol = 1e-10){
               apply(cbind(data, data.sg), 1, fun.for.integrate,
                     deg = deg, degree = degree, x.max = x.max, x.min = x.min,
                     log = log, abs.tol = abs.tol)},
             data = R.test, data.sg = R.sg.test, deg = deg, x.max = R.max, x.min = R.min,
             log = log, abs.tol = abs.tol)
    R.indv.pdf <- as.matrix(res.R.pdf, n.test, deg)
  }
  
  if (n.test == 1) {
    C.pdf <- kronecker(M.indv.pdf, R.indv.pdf)
  } else {
    C.pdf <- sapply(1:n.test, function(x, y, i){kronecker(x[i,], y[i,])}, 
                    x = M.indv.pdf, y = R.indv.pdf)
  }
  
  loglike.pred <- sum(log(weights %*% C.pdf))
  
  return(loglike.pred)
}

File: PyCode/Cross_Validation.py
import numpy as np
from scipy.stats import beta,norm
from scipy.integrate import quad
from scipy.optimize import brentq as root
from astropy.table import Table
import datetime,os
import sys 
sys.path.append(os.path.dirname(__file__))
from MLE_fit import MLE_fit, find_indv_pdf, integrate_function
  
  
#' @param data.train: input training data
#' @param data.sg.train: input measurement errors for the training data
#' @param bounds: a vector contains four elements, from left to right:
#'                the upper bound for mass, the lower bound for mass
#'                the upper bound for radius, the lower bound for radius
#' @param data.test: input test data
#' @param data.sg.test: input measurement errors for the test data
#' @param deg: degree used in Bernstein polynomials
#' @param abs.tol: tolerance number of computing


def cross_validation(train_data, bounds, test_radius, test_mass, deg, train_data_sg = None, test_mass_sigma = None, test_radius_sigma = None, Log = False, abs_tol = 1e-10, location = os.path.dirname(__file__)):
    '''
    INPUT:
        train_data = Input training data
        bounds = Vecotr containing four elements. Upper and lower bound for Mass. Upper and lower bound for Radius resp.
        test_radius, = Input radius data
        test_mass, = Input mass data
        deg = Degree used in Bernstein Polynomials
        train_data_sg = Input training data measurement errors
        test_radius_sigma = Input test radius measurement errors
        test_mass_sigma = Input test mass measurement errors
        Log = Whether the data is in log units
        abs_tol = Absolute tolerance       
        location : The location for the log file
    ''' 
    
    if np.shape(test_radius_sigma) != np.shape(test_mass_sigma):
        print('Insert error message here')
    elif np.shape(test_radius) != np.shape(test_mass):
        print('Gibberish')
    
    # Fit MLE using training dataset
    weights = MLE_fit(data = train_data, bounds = bounds, sigma = train_data_sg, Log = Log, deg = deg, abs_tol = abs_tol, location = location)
    
    size_test = np.size(test_radius)
    
    # specify the bounds
    M_max = bounds[1]
    M_min = bounds[2]
    R_max = bounds[3]
    R_min = bounds[4]
    
    # calculate cdf and pdf of M and R for each term
    # the first and last term is set to 0 to avoid boundary effects
    # so we only need to calculate 2:(deg^2-1) terms
    deg_vec = np.arange(1,deg+1) 
    
    if test_radius_sigma is None:
        # PDF for Mass and Radius
        M_indv_pdf = find_indv_pdf(test_mass,deg,deg_vec,M_max,M_min) 
        R_indv_pdf = find_indv_pdf(test_radius,deg,deg_vec,R_max,R_min) 
    else:
        # PDF for Integrated beta density and normal density for Mass and Radius
        M_indv_pdf = np.array([integrate_function(data = test_mass, data_sd = test_mass_sigma, deg = deg, degree = d, x_max = M_max, x_min = M_min) for d in deg_vec])        
        R_indv_pdf = np.array([integrate_function(data = test_radius, data_sd = test_radius_sigma, deg = deg, degree = d, x_max = R_max, x_min = R_min) for d in deg_vec])  
        
        


    
    

